File: python/indicators.py
"""
Indicator calculations using pandas_ta.
Supports all 57 indicators from the Flowchart app.
"""

import pandas as pd
import numpy as np
import pandas_ta as ta

def calculate_indicator(df: pd.DataFrame, indicator_id: str, period: int = None):
    """
    Calculate an indicator and return the series.

    Args:
        df: DataFrame with OHLCV columns (Date, Open, High, Low, Close, Adj Close, Volume)
        indicator_id: Indicator name (e.g., 'RSI', 'SMA')
        period: Window period (ignored for windowless indicators)

    Returns:
        pd.Series with indicator values
    """
    # Use Adj Close for calculations (accounts for splits/dividends)
    close = df['Adj Close'] if 'Adj Close' in df.columns else df['Close']
    high = df['High']
    low = df['Low']
    open_ = df['Open']
    volume = df['Volume']

    # ============================================================================
    # PRICE (1)
    # ============================================================================
    if indicator_id == 'Current Price':
        return close

    # ============================================================================
    # MOVING AVERAGES (8)
    # ============================================================================
    elif indicator_id == 'Simple Moving Average':
        return ta.sma(close, length=period)

    elif indicator_id == 'Exponential Moving Average':
        return ta.ema(close, length=period)

    elif indicator_id == 'Hull Moving Average':
        return ta.hma(close, length=period)

    elif indicator_id == 'Weighted Moving Average':
        return ta.wma(close, length=period)

    elif indicator_id == 'Wilder Moving Average':
        return ta.rma(close, length=period)  # RMA = Wilder smoothing

    elif indicator_id == 'DEMA':
        return ta.dema(close, length=period)

    elif indicator_id == 'TEMA':
        return ta.tema(close, length=period)

    elif indicator_id == 'KAMA':
        return ta.kama(close, length=period)

    # ============================================================================
    # RSI & VARIANTS (5)
    # ============================================================================
    elif indicator_id == 'RSI':
        return ta.rsi(close, length=period)

    elif indicator_id == 'RSI (SMA)':
        # RSI with SMA smoothing instead of Wilder
        gain = close.diff()
        avg_gain = gain.where(gain > 0, 0).rolling(window=period).mean()
        avg_loss = -gain.where(gain < 0, 0).rolling(window=period).mean()
        rs = avg_gain / avg_loss
        return 100 - (100 / (1 + rs))

    elif indicator_id == 'RSI (EMA)':
        # RSI with EMA smoothing
        gain = close.diff()
        avg_gain = gain.where(gain > 0, 0).ewm(span=period, adjust=False).mean()
        avg_loss = -gain.where(gain < 0, 0).ewm(span=period, adjust=False).mean()
        rs = avg_gain / avg_loss
        return 100 - (100 / (1 + rs))

    elif indicator_id == 'Stochastic RSI':
        return ta.stochrsi(close, length=period)['STOCHRSIk_14_14_3_3']  # %K line

    elif indicator_id == 'Laguerre RSI':
        # Laguerre RSI (simplified implementation)
        gamma = 0.5
        L0 = close.copy()
        L1 = L0.shift(1)
        L2 = L1.shift(1)
        L3 = L2.shift(1)

        for i in range(len(close)):
            if i >= 3:
                L0.iloc[i] = (1 - gamma) * close.iloc[i] + gamma * L0.iloc[i-1]
                L1.iloc[i] = -gamma * L0.iloc[i] + L0.iloc[i-1] + gamma * L1.iloc[i-1]
                L2.iloc[i] = -gamma * L1.iloc[i] + L1.iloc[i-1] + gamma * L2.iloc[i-1]
                L3.iloc[i] = -gamma * L2.iloc[i] + L2.iloc[i-1] + gamma * L3.iloc[i-1]

        CU = pd.Series(0.0, index=close.index)
        CD = pd.Series(0.0, index=close.index)

        for i in range(len(close)):
            if L0.iloc[i] >= L1.iloc[i]:
                CU.iloc[i] = L0.iloc[i] - L1.iloc[i]
            else:
                CD.iloc[i] = L1.iloc[i] - L0.iloc[i]

            if L1.iloc[i] >= L2.iloc[i]:
                CU.iloc[i] += L1.iloc[i] - L2.iloc[i]
            else:
                CD.iloc[i] += L2.iloc[i] - L1.iloc[i]

            if L2.iloc[i] >= L3.iloc[i]:
                CU.iloc[i] += L2.iloc[i] - L3.iloc[i]
            else:
                CD.iloc[i] += L3.iloc[i] - L2.iloc[i]

        lrsi = CU / (CU + CD) * 100
        return lrsi

    # ============================================================================
    # MOMENTUM (9)
    # ============================================================================
    elif indicator_id == 'Momentum (Weighted)':
        # 1-3-6-12 weighted momentum
        mom1 = close.pct_change(1)
        mom3 = close.pct_change(3)
        mom6 = close.pct_change(6)
        mom12 = close.pct_change(12)
        return (mom1 * 12 + mom3 * 4 + mom6 * 2 + mom12) / 19 * 100

    elif indicator_id == 'Momentum (Unweighted)':
        # 1-3-6-12 unweighted momentum
        mom1 = close.pct_change(1)
        mom3 = close.pct_change(3)
        mom6 = close.pct_change(6)
        mom12 = close.pct_change(12)
        return (mom1 + mom3 + mom6 + mom12) / 4 * 100

    elif indicator_id == 'Momentum (12-Month SMA)':
        return ta.sma(close.pct_change(period), length=12) * 100

    elif indicator_id == 'Rate of Change':
        return ta.roc(close, length=period)

    elif indicator_id == 'Williams %R':
        return ta.willr(high, low, close, length=period)

    elif indicator_id == 'CCI':
        return ta.cci(high, low, close, length=period)

    elif indicator_id == 'Stochastic %K':
        stoch = ta.stoch(high, low, close, k=period, d=3)
        return stoch[f'STOCHk_{period}_3_3']

    elif indicator_id == 'Stochastic %D':
        stoch = ta.stoch(high, low, close, k=period, d=3)
        return stoch[f'STOCHd_{period}_3_3']

    elif indicator_id == 'ADX':
        adx = ta.adx(high, low, close, length=period)
        return adx[f'ADX_{period}']

    # ============================================================================
    # VOLATILITY (10)
    # ============================================================================
    elif indicator_id == 'Standard Deviation':
        returns = close.pct_change()
        return returns.rolling(window=period).std() * np.sqrt(252) * 100

    elif indicator_id == 'Standard Deviation of Price':
        return close.rolling(window=period).std()

    elif indicator_id == 'Max Drawdown':
        # Maximum drawdown over period
        rolling_max = close.rolling(window=period, min_periods=1).max()
        drawdown = (close - rolling_max) / rolling_max * 100
        return drawdown.rolling(window=period).min()

    elif indicator_id == 'Drawdown':
        # Current drawdown from all-time high within period
        rolling_max = close.rolling(window=period, min_periods=1).max()
        return (close - rolling_max) / rolling_max * 100

    elif indicator_id == 'Bollinger %B':
        bb = ta.bbands(close, length=period, std=2)
        lower = bb[f'BBL_{period}_2.0']
        upper = bb[f'BBU_{period}_2.0']
        return (close - lower) / (upper - lower)

    elif indicator_id == 'Bollinger Bandwidth':
        bb = ta.bbands(close, length=period, std=2)
        lower = bb[f'BBL_{period}_2.0']
        upper = bb[f'BBU_{period}_2.0']
        middle = bb[f'BBM_{period}_2.0']
        return ((upper - lower) / middle) * 100

    elif indicator_id == 'ATR':
        return ta.atr(high, low, close, length=period)

    elif indicator_id == 'ATR %':
        atr = ta.atr(high, low, close, length=period)
        return (atr / close) * 100

    elif indicator_id == 'Historical Volatility':
        returns = np.log(close / close.shift(1))
        return returns.rolling(window=period).std() * np.sqrt(252) * 100

    elif indicator_id == 'Ulcer Index':
        # Ulcer Index = sqrt(mean((% drawdown)^2))
        rolling_max = close.rolling(window=period, min_periods=1).max()
        drawdown_pct = ((close - rolling_max) / rolling_max) * 100
        return np.sqrt((drawdown_pct ** 2).rolling(window=period).mean())

    # ============================================================================
    # TREND (7)
    # ============================================================================
    elif indicator_id == 'Cumulative Return':
        return (close / close.shift(period) - 1) * 100

    elif indicator_id == 'SMA of Returns':
        returns = close.pct_change()
        return ta.sma(returns, length=period) * 100

    elif indicator_id == 'Trend Clarity':
        # R-squared of linear regression
        def rsquared(y):
            if len(y) < 2:
                return np.nan
            x = np.arange(len(y))
            mask = ~np.isnan(y)
            if mask.sum() < 2:
                return np.nan
            slope, intercept = np.polyfit(x[mask], y[mask], 1)
            y_pred = slope * x + intercept
            ss_res = np.sum((y - y_pred) ** 2)
            ss_tot = np.sum((y - np.mean(y)) ** 2)
            return 1 - (ss_res / ss_tot) if ss_tot != 0 else np.nan

        return close.rolling(window=period).apply(rsquared, raw=True)

    elif indicator_id == 'Ultimate Smoother':
        # Ehlers Ultimate Smoother (simplified)
        a = 2 / (period + 1)
        smoothed = close.ewm(alpha=a, adjust=False).mean()
        return smoothed

    elif indicator_id == 'Linear Reg Slope':
        def slope(y):
            if len(y) < 2:
                return np.nan
            x = np.arange(len(y))
            mask = ~np.isnan(y)
            if mask.sum() < 2:
                return np.nan
            return np.polyfit(x[mask], y[mask], 1)[0]

        return close.rolling(window=period).apply(slope, raw=True)

    elif indicator_id == 'Linear Reg Value':
        return ta.linreg(close, length=period)

    elif indicator_id == 'Price vs SMA':
        sma = ta.sma(close, length=period)
        return close / sma

    # ============================================================================
    # AROON (3)
    # ============================================================================
    elif indicator_id == 'Aroon Up':
        aroon = ta.aroon(high, low, length=period)
        return aroon[f'AROONU_{period}']

    elif indicator_id == 'Aroon Down':
        aroon = ta.aroon(high, low, length=period)
        return aroon[f'AROOND_{period}']

    elif indicator_id == 'Aroon Oscillator':
        aroon = ta.aroon(high, low, length=period)
        return aroon[f'AROONOSC_{period}']

    # ============================================================================
    # MACD/PPO (2)
    # ============================================================================
    elif indicator_id == 'MACD Histogram':
        macd = ta.macd(close, fast=12, slow=26, signal=9)
        return macd['MACDh_12_26_9']

    elif indicator_id == 'PPO Histogram':
        ppo = ta.ppo(close, fast=12, slow=26, signal=9)
        return ppo['PPOh_12_26_9']

    # ============================================================================
    # VOLUME (3)
    # ============================================================================
    elif indicator_id == 'Money Flow Index':
        return ta.mfi(high, low, close, volume, length=period)

    elif indicator_id == 'OBV Rate of Change':
        obv = ta.obv(close, volume)
        return obv.pct_change(period) * 100

    elif indicator_id == 'VWAP Ratio':
        vwap = ta.vwap(high, low, close, volume, anchor='D')
        return close / vwap

    else:
        raise ValueError(f"Unknown indicator: {indicator_id}")


def calculate_all_periods(df: pd.DataFrame, indicator_id: str, periods: list):
    """
    Calculate an indicator for multiple periods at once (optimization).

    Args:
        df: DataFrame with OHLCV columns
        indicator_id: Indicator name
        periods: List of periods to calculate

    Returns:
        dict: {period: pd.Series}
    """
    results = {}
    for period in periods:
        results[period] = calculate_indicator(df, indicator_id, period)
    return results


if __name__ == '__main__':
    # Test with sample data
    import yfinance as yf

    print("Testing indicator calculations...")
    data = yf.download('SPY', start='2020-01-01', end='2024-01-01', progress=False)
    data = data.reset_index()
    data.columns = ['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume']

    # Test RSI
    rsi = calculate_indicator(data, 'RSI', period=14)
    print(f"RSI(14) last 5 values:\n{rsi.tail()}")

    # Test SMA
    sma = calculate_indicator(data, 'Simple Moving Average', period=50)
    print(f"\nSMA(50) last 5 values:\n{sma.tail()}")

    print("\nIndicator calculations working correctly!")

File: app/src/constants/indicators.ts
// src/constants/indicators.ts
// Indicator configuration and metadata

import type { MetricChoice } from '../types'

// Indicators that don't require a window parameter
export const WINDOWLESS_INDICATORS: MetricChoice[] = [
  'Current Price',
  'Date',                     // Calendar-based, no lookback needed
  'Momentum (Weighted)',      // Fixed 1/3/6/12 month
  'Momentum (Unweighted)',    // Fixed 1/3/6/12 month
  'Momentum (12-Month SMA)',  // Fixed 12 month
  'Drawdown',                 // Uses all history from ATH
  'MACD Histogram',           // Fixed 12/26/9
  'PPO Histogram',            // Fixed 12/26/9
  'Laguerre RSI',             // Uses gamma instead of window
]

export const isWindowlessIndicator = (metric: MetricChoice): boolean =>
  WINDOWLESS_INDICATORS.includes(metric)

// Get the actual lookback period needed for an indicator (for warm-up calculations)
export const getIndicatorLookback = (metric: MetricChoice, window: number): number => {
  switch (metric) {
    case 'Current Price':
    case 'Date':
      return 0
    case 'Momentum (Weighted)':
    case 'Momentum (Unweighted)':
      return 252 // 12 months of trading days
    case 'Momentum (12-Month SMA)':
      return 252 // 12 months
    case 'Drawdown':
      return 0 // Uses all history, no fixed lookback needed for warm-up
    case 'MACD Histogram':
    case 'PPO Histogram':
      return 35 // 26 + 9 for signal line
    case 'Laguerre RSI':
      return 10 // Minimal lookback, uses recursive filter
    case 'DEMA':
      return Math.max(1, Math.floor(window || 0)) * 2
    case 'TEMA':
      return Math.max(1, Math.floor(window || 0)) * 3
    case 'KAMA':
      return Math.max(1, Math.floor(window || 0)) + 30 // window + slow period
    case 'ADX':
      return Math.max(1, Math.floor(window || 0)) * 2 // Extra for smoothing
    default:
      return Math.max(1, Math.floor(window || 0))
  }
}

// Indicator metadata with descriptions and formulas
export const INDICATOR_INFO: Record<MetricChoice, { desc: string; formula: string }> = {
  // Price
  'Current Price': { desc: 'The current closing price of the asset', formula: 'Price = Close' },

  // Date-based
  'Date': { desc: 'Check if current date falls within a range (for seasonal strategies)', formula: 'True if dateFrom <= currentDate <= dateTo' },

  // Moving Averages
  'Simple Moving Average': { desc: 'Average price over N periods, equal weight to all', formula: 'SMA = Σ(Close) / N' },
  'Exponential Moving Average': { desc: 'Weighted average giving more weight to recent prices', formula: 'EMA = α×Close + (1-α)×EMA_prev, α=2/(N+1)' },
  'Hull Moving Average': { desc: 'Reduced lag MA using weighted MAs', formula: 'HMA = WMA(2×WMA(N/2) - WMA(N), √N)' },
  'Weighted Moving Average': { desc: 'Linear weighted average, recent prices weighted more', formula: 'WMA = Σ(i×Close_i) / Σ(i), i=1..N' },
  'Wilder Moving Average': { desc: 'Smoothed MA used in RSI, slower response', formula: 'WilderMA = (Prev×(N-1) + Close) / N' },
  'DEMA': { desc: 'Double EMA reduces lag vs single EMA', formula: 'DEMA = 2×EMA - EMA(EMA)' },
  'TEMA': { desc: 'Triple EMA for even less lag', formula: 'TEMA = 3×EMA - 3×EMA² + EMA³' },
  'KAMA': { desc: 'Adapts smoothing based on market noise', formula: 'KAMA = KAMA_prev + SC²×(Close - KAMA_prev)' },

  // RSI & Variants
  'Relative Strength Index': { desc: "Wilder's RSI, momentum oscillator 0-100", formula: 'RSI = 100 - 100/(1 + AvgGain/AvgLoss)' },
  'RSI (SMA)': { desc: 'RSI using simple moving average smoothing', formula: 'RSI_SMA = 100 - 100/(1 + SMA(Gains)/SMA(Losses))' },
  'RSI (EMA)': { desc: 'RSI using exponential moving average smoothing', formula: 'RSI_EMA = 100 - 100/(1 + EMA(Gains)/EMA(Losses))' },
  'Stochastic RSI': { desc: 'Stochastic applied to RSI values, more sensitive', formula: 'StochRSI = (RSI - RSI_Low) / (RSI_High - RSI_Low)' },
  'Laguerre RSI': { desc: "Ehlers' Laguerre filter RSI, smoother", formula: 'Uses 4-element Laguerre filter with gamma' },

  // Momentum
  'Momentum (Weighted)': { desc: '13612W weighted momentum score', formula: '12×M1 + 4×M3 + 2×M6 + M12 (1/3/6/12 month returns)' },
  'Momentum (Unweighted)': { desc: '13612U unweighted momentum score', formula: 'M1 + M3 + M6 + M12 (equal weight)' },
  'Momentum (12-Month SMA)': { desc: 'SMA of 12-month returns', formula: 'SMA(12-month return, N)' },
  'Rate of Change': { desc: 'Percent change over N periods', formula: 'ROC = (Close - Close_N) / Close_N × 100' },
  'Williams %R': { desc: 'Momentum indicator, inverse of Fast Stochastic', formula: '%R = (High_N - Close) / (High_N - Low_N) × -100' },
  'CCI': { desc: 'Measures price deviation from average', formula: 'CCI = (TP - SMA(TP)) / (0.015 × MeanDev)' },
  'Stochastic %K': { desc: 'Fast Stochastic, price position in range', formula: '%K = (Close - Low_N) / (High_N - Low_N) × 100' },
  'Stochastic %D': { desc: 'Slow Stochastic, SMA of %K', formula: '%D = SMA(%K, 3)' },
  'ADX': { desc: 'Trend strength indicator 0-100', formula: 'ADX = SMA(|+DI - -DI| / (+DI + -DI) × 100)' },

  // Volatility
  'Standard Deviation': { desc: 'Volatility of returns over N periods', formula: 'StdDev = √(Σ(r - r̄)² / N)' },
  'Standard Deviation of Price': { desc: 'Volatility of price levels (absolute $)', formula: 'StdDev = √(Σ(P - P̄)² / N)' },
  'Max Drawdown': { desc: 'Largest peak-to-trough decline over N periods', formula: 'MaxDD = max((Peak - Trough) / Peak)' },
  'Drawdown': { desc: 'Current decline from recent peak', formula: 'DD = (Peak - Current) / Peak' },
  'Bollinger %B': { desc: 'Position within Bollinger Bands (0-1)', formula: '%B = (Close - LowerBand) / (UpperBand - LowerBand)' },
  'Bollinger Bandwidth': { desc: 'Width of Bollinger Bands as % of middle', formula: 'BW = (Upper - Lower) / Middle × 100' },
  'ATR': { desc: 'Average True Range, volatility measure', formula: 'ATR = SMA(max(H-L, |H-C_prev|, |L-C_prev|))' },
  'ATR %': { desc: 'ATR as percentage of price', formula: 'ATR% = ATR / Close × 100' },
  'Historical Volatility': { desc: 'Annualized standard deviation of returns', formula: 'HV = StdDev(returns) × √252' },
  'Ulcer Index': { desc: 'Measures downside volatility/drawdown pain', formula: 'UI = √(Σ(DD²) / N)' },

  // Trend
  'Cumulative Return': { desc: 'Total return over N periods', formula: 'CumRet = (Close / Close_N) - 1' },
  'SMA of Returns': { desc: 'Smoothed average of daily returns', formula: 'SMA(daily returns, N)' },
  'Trend Clarity': { desc: 'R² of price regression, trend strength', formula: 'R² = 1 - (SS_res / SS_tot)' },
  'Ultimate Smoother': { desc: "Ehlers' low-lag smoother", formula: '3-pole Butterworth filter' },
  'Linear Reg Slope': { desc: 'Slope of best-fit line through prices', formula: 'Slope = Σ((x-x̄)(y-ȳ)) / Σ(x-x̄)²' },
  'Linear Reg Value': { desc: 'Current value on regression line', formula: 'Value = Intercept + Slope × N' },
  'Price vs SMA': { desc: 'Ratio of price to its moving average', formula: 'Ratio = Close / SMA(Close, N)' },

  // Aroon
  'Aroon Up': { desc: 'Days since highest high (0-100)', formula: 'AroonUp = ((N - DaysSinceHigh) / N) × 100' },
  'Aroon Down': { desc: 'Days since lowest low (0-100)', formula: 'AroonDown = ((N - DaysSinceLow) / N) × 100' },
  'Aroon Oscillator': { desc: 'Difference between Aroon Up and Down', formula: 'AroonOsc = AroonUp - AroonDown' },

  // MACD/PPO
  'MACD Histogram': { desc: 'MACD minus signal line', formula: 'Hist = (EMA12 - EMA26) - EMA9(EMA12 - EMA26)' },
  'PPO Histogram': { desc: 'Percentage Price Oscillator histogram', formula: 'PPO = ((EMA12 - EMA26) / EMA26) × 100' },

  // Volume-based
  'Money Flow Index': { desc: 'Volume-weighted RSI, measures buying/selling pressure', formula: 'MFI = 100 - 100/(1 + PosMF/NegMF)' },
  'OBV Rate of Change': { desc: 'Momentum of cumulative On-Balance Volume', formula: 'OBV ROC = (OBV - OBV_N) / |OBV_N| × 100' },
  'VWAP Ratio': { desc: 'Price vs Volume-Weighted Avg Price (100 = at VWAP)', formula: 'Ratio = Close / VWAP × 100' },
}

// Indicator categories for submenu dropdown
export const INDICATOR_CATEGORIES: Record<string, MetricChoice[]> = {
  'Price': ['Current Price'],
  'Date/Calendar': ['Date'],
  'Moving Averages': [
    'Simple Moving Average',
    'Exponential Moving Average',
    'Hull Moving Average',
    'Weighted Moving Average',
    'Wilder Moving Average',
    'DEMA',
    'TEMA',
    'KAMA',
  ],
  'RSI & Variants': [
    'Relative Strength Index',
    'RSI (SMA)',
    'RSI (EMA)',
    'Stochastic RSI',
    'Laguerre RSI',
  ],
  'Momentum': [
    'Momentum (Weighted)',
    'Momentum (Unweighted)',
    'Momentum (12-Month SMA)',
    'Rate of Change',
    'Williams %R',
    'CCI',
    'Stochastic %K',
    'Stochastic %D',
    'ADX',
  ],
  'Volatility': [
    'Standard Deviation',
    'Standard Deviation of Price',
    'Max Drawdown',
    'Drawdown',
    'Bollinger %B',
    'Bollinger Bandwidth',
    'ATR',
    'ATR %',
    'Historical Volatility',
    'Ulcer Index',
  ],
  'Trend': [
    'Cumulative Return',
    'SMA of Returns',
    'Trend Clarity',
    'Ultimate Smoother',
    'Linear Reg Slope',
    'Linear Reg Value',
    'Price vs SMA',
  ],
  'Aroon': [
    'Aroon Up',
    'Aroon Down',
    'Aroon Oscillator',
  ],
  'MACD/PPO': [
    'MACD Histogram',
    'PPO Histogram',
  ],
  'Volume': [
    'Money Flow Index',
    'OBV Rate of Change',
    'VWAP Ratio',
  ],
}
